### PR TITLE
Si 2712 enhancement for recursive higher-kinded structures

### PR DESCRIPTION
**Pull request by mandubian**
With current patch,

you can unify 

```
F[T] <=> Bar[Int, String, T]
F[T] <=> (Int => T)
```

but not:

```
F[T] <=> Option[String, Bar[Int, String, T]]
F[T] <=> Option[List[T], Bar[Int, String, T]]
```

At first view, it should be possible (using the same left-right precedences rules)

This patch is aimed at solving this issue by introspecting recursively higher-kinded structures in depth trying to unify every internal structure with `F[T]`.

It's just another case tried when basic solution doesn't work so all existing tests work.
And it allows to compile straight-forward this kind of code:

```
// Shapeless Coproduct
sealed trait Coproduct
@scala.annotation.unifyRightToLeft
sealed trait :+:[+H, +T <: Coproduct] extends Coproduct
sealed trait CNil extends Coproduct

// cats/scalaz TransNat
trait NaturalTransformation[F[_], G[_]]
type ~>[F[_], G[_]] = NaturalTransformation[F, G]

// a function using trans-nat (like fold in Free)
def foo[F[_], G[_]](f: F ~> G): Unit = ???

type Id[A] = A

trait Bar[A, B, C]
val a: ({type λ[T] = Bar[Int, String, T] })#λ ~> Id = ???
foo(a)

val b: ({type λ[T] = T :+: CNil })#λ ~> Id = ???
foo(b)

val c : ({type λ[T] = Option[T] :+: String :+: List[T] :+: Bar[Int, String, T] :+: CNil })#λ ~> Id = ???
foo(c)
```

... Which might change a lot how we can manipulate multi-higher-kinded structures like Free or Coproducts in our libs and code...

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -3154,39 +3190,87 @@ trait Types
             //
             // A more "natural" unifier might be M[t] = [t][t => t]. There's lots of scope for
             // experimenting with alternatives here.
-
             val tpSym = tp.typeSymbolDirect
-            val rightToLeft = tpSym.hasAnnotation(definitions.unifyRightToLeftClass)
+            val rightToLeft = tpSym.annotations.exists(_ matches definitions.unifyRightToLeftClass)
```
Comment (milessabin) [inline src/reflect/scala/reflect/internal/Types.scala]: What's going on here? It looks like you blown away some of the improvements that Jason made to my original version?

Comment (mandubian) [inline src/reflect/scala/reflect/internal/Types.scala]: argllll... sorry, correcting that

**Comment by milessabin** [comment]
Can you say how the original test case unifies? Is it `A` = `Int` and `M` = `[t][t => t]`?

**Comment by milessabin** [comment]
I'm afraid I'm not getting much clearer on what's going on by reading the code. Could you write down the algorithm abstractly?

**Comment by mandubian** [comment]
Have you seen this Gist?
https://gist.github.com/mandubian/7e46212957aed8b0d132616c62488b08

**Comment by mandubian** [comment]
Have a look at this test

```
def foo3[F[_], A](t: F[A])(implicit eq: F[A] =:= ({type λ[T] = T => T })#λ[A]): F[A] = t
def f[T]: T => T = ???
val t0 = foo3(f)
```

and it fails as:

```
t2712-11.scala:34: error: Cannot prove that Nothing => A =:= A => A.
  val t0 = foo3(f)
               ^
one error found
```

So it is `[t][Int => t]` and it seems normal IMHO as I just re-use your original algorithm on every higher-kinded type arguments in higher-kinded types...

`M[F1[...], F2[...]]`
Just applies your left-to-right/right-to-lefet algorithm on `F1` and `F2` and if both are unifiable, build `M[T]`
it just adds the ability to unify in depth in higher-kinded structures, nothing else

**Comment by mandubian** [comment]
and that compiles:

```
  def foo3[F[_], A](t: F[A])(implicit eq: F[A] =:= ({type λ[T] = Int => T })#λ[A]): F[A] = t
  val t1 = foo3{(x: Int) => x}
```

**Comment by milessabin** [comment]
I've looked at the gist, but that's not really what I meant by "abstractly" ... you're working through a single concrete example.

**Comment by mandubian** [comment]
I can write it abstractly, this gist was written yesterday just to help understanding

**Comment by milessabin** [comment]
Using the implicit argument in the way you have in `foo3` is problematic because it could affect type inference itself. Try something like,

``` scala
def typed[T](t: T): Unit = {}

def meh[F[_], A](t: F[A]): F[Boolean] = ???
val res = meh{(x: Int) => x}

typed[Boolean => Boolean](res)
```

In my PR the last line should fail because the inferred type will be `Int => Boolean`. What happens with your variant?

**Comment by mandubian** [comment]
yes you're right, this one is better

```
t2712-11.scala:37: error: type mismatch;
 found   : Int => Boolean
 required: Boolean => Boolean
  typed[Boolean => Boolean](res)
                            ^
one error found
```

**Comment by milessabin** [comment]
Sorry I don't understand ... is that the result with your version of the algorithm?

**Comment by mandubian** [comment]
Yes it is my algorithm.
Written as it is for now, it doesn't change anything to your algo for this case

**Comment by milessabin** [comment]
I think that's a problem. It really _should_ change the result in this case. If it doesn't then I'm afraid I have no solid intuitions about how it would behave in more complex cases.

**Comment by mandubian** [comment]
Let me try to write the algo abstractly to help you think...
In my head, it's clear how it woks: it applies your algorithm for every typeargs recursively if and only if there are higher-kind type-args in type args.
But I don't say it's right, it's just naive, your vision with [t](t => t) seems quite logical...

**Comment by milessabin** [comment]
OK, but I don't think that's the right way to go.

The simple algorithm has the merit of being simple and easy to understand, but produces sub-optimal unifications because it's so simple minded, hence `[t](Int => t)`.

If we're going to have something more complex, then it should at least deliver the intuitive unification `[t](t => t)` in this case.

**Comment by mandubian** [comment]
Yes I agree... and as you said yesterday we should be able to go to simpler case with a type alias.
Recursion in depth as I implemented is useful anyway IMHO.
Let's try to enhance this algorithm for our better goal...

**Comment by milessabin** [comment]
:+1:

Writing up an abstract description would be a good first step.

**Comment by mandubian** [comment]
## Recursive Unification Algorithm = function unifySpecific

### Inputs
- local type (with typeArgs)
- `tp`: type to be unified with local type
- `rootLevel`: a flag to know if we are on higher-level of tp (ie not in a recursion loop)
- `freeSyms`: current free symbols built from local type (in recursion loops, we need to re-use the same free symbols when building types)

### Returns
- isUnified : a flag indicating whether tp is unifiable with local type
- type : in case of recursion, this value is the unified built type using current freeSyms

### Algo

`// Classic case`
- if (typeArgs.length == tp.typeArgs.length) `// same length so trivial unification`
  - check if tp can be unified (withinbounds && isSubArgs)
    - if true
      - if (rootLevel)
        - addBound(tp.typeConstructor)
        - return (true, tp) `// returning tp is arbitrary as this value is not used at rootLevel`
      - else `// we are in a recursion loop so build type using passed freeSyms`
        - build appliedType(tp.typeConstructor, freesyms)
        - return (true, appliedType)
    - else return (false, tp) `// arbitrary returning tp as not used`

`// SI2712 patch`
- if (settings.YhigherOrderUnification && 0 < typeArgs < tp.typeArgs) `tp has more args than local type`
  - build localFreeSyms
    - if (rootLevel) build root freeSymbols from local typeParams (depending on left/right order)
    - else re-use passed freeSyms `// in recursion, the same freesyms must be reused for all built types`
  - check if any tp.typeargs is higher-kinded
    - if false `// original SI2712 patch algo`
      - extract (captured, abstracted) typeArgs from local type (depending on left/right order)
      - check if tp can be unified (withinbounds && isSubArgs)
        - if(unifiable)
          - build appliedType(tp.typeConstructor, localFreeSyms ++ captured) (depending on left/right order)
          - if (rootLevel) addBound(PolyType(localFreeSyms, appliedType))
          - return (true, appliedType)
        - else return (false, tp) `// arbitrary tp not used`
    - else `// we need to try unifying in depth by recursing on HK typeargs`
      - for every tp.typeArgs : tpp
        - if tpp is higher-kinded
          `// HERE IS THE RECURSION`
          - call recursion on unifySpecific(tpp, not(RootLevel), localFreeSyms) => (isUnifiable, typeUnified)
            - if (not isUnifiable)
              - stop going through tp.typeArgs
            - else 
              - keep typeUnified in unifiableTypes
              - go to next tp.typeArgs
        - else
          `// this type is simple-kinded so is not subject to unification but is needed to rebuild type with localFreeSyms`
          - keep the type in unifiableTypes
      - check that all tp.typeArgs were unifiable && NOT all simple-kinded
        - if true
          - build appliedType(tp.typeConstructor, unifiableTypes)
          - if (rootLevel) addBound(PolyType(localFreeSyms, appliedType))
          - return (true, appliedType)
        - else
          - return (false, tp) `// arbitrary tp not used`
- else return (false, tp) `// arbitrary tp not used`

**Pull request closed without merging**